lib/src/facts/bsd/networking_resolver.cc

```diff
@@ -153,6 +156,56 @@ namespace facter { namespace facts { namespace bsd {
                 return true;
             }, "^dhclient.*lease.*$");
         }
+    }
+
+    void networking_resolver::find_networkd_dhcp_servers(std::map<std::string, std::string>& servers) const
+    {
+        static const string networkd_lease_directory = "/run/systemd/netif/leases/";
+
+        if (!boost::filesystem::is_directory(networkd_lease_directory)) return;
+
+        static boost::regex ip_link_re("^(\\d+):\\s+([^:]+)");
+        unordered_map<int, string> iface_index_names;
+        string key, value;
+
+        // Gather a map of interface indices and their interface names from `ip link show`
+        // Only the index is known in the lease files
+        each_line("ip", {"link", "show"}, [&](string line) {
+            if (re_search(line, ip_link_re, &key, &value)) {
+                iface_index_names.insert(make_pair(stoi(key), value));
+            }
+            return true;
+        });
+
+        LOG_DEBUG("searching \"{1}\" for systemd-networkd DHCP lease files", networkd_lease_directory);
+
+        lth_file::each_file(networkd_lease_directory, [&](string const& path) {
+            LOG_DEBUG("searching \"{1}\" for systemd-networkd DHCP lease information", path);
+            string server_address;
+
+            static boost::regex server_address_re("^SERVER_ADDRESS=(.*)$");
+            lth_file::each_line(path, [&](string& line) {
+                boost::trim(line);
+                if (re_search(line, server_address_re, &server_address)) {
+                    boost::filesystem::path p {path};
+                    // systemd-networkd lease files are named after the interface index
+                    auto iface_index = maybe_stoi(p.filename().string());
+                    if (!iface_index) return true;
+                    servers.emplace(make_pair(iface_index_names[iface_index.get()], server_address));
+                }
+                return true;
+            });
+            return true;
+        });
+    }
+
+    map<string, string> networking_resolver::find_dhcp_servers() const
+    {
+        map<string, string> servers;
+
+        find_networkd_dhcp_servers(servers);
+        find_dhclient_dhcp_servers(servers);
```

Comment:
I think instead of trying to do both, we should figure out which file exists and explicitly choose the more-correct one.